### PR TITLE
Refactoring and tests

### PR DESCRIPTION
**Pull request by FredFaust**

### CODE REVIEW

File: dist/marionette-tree-view.js [hunk 1]
```diff
@@ -13,129 +13,155 @@ Marionette.NodeView = Marionette.CompositeView.extend({
...
+    'change:children': 'onChangeChildren',
+    'delete:children': 'onDeleteChildren',
+    'error': 'onError'
   },
 
   initialize: function(options) {
+    var models = [];
+
     options = options || {};
+    this.nbChildrenAttrName = ((options.nbChildrenAttrName) ? options.nbChildrenAttrName : 'nbChildren');
```
Comment (Nelrohd) [inline dist/marionette-tree-view.js]: why parentheses around options.nbChildrenAttrName?

File: test/unit/spec/node.js [hunk 2]
```diff
@@ -1,115 +1,196 @@
...
-
-    expect(customNodeView.onExpand).toHaveBeenCalled();
-  });
+  var CustomNodeView = Marionette.NodeView.extend({
+        template: _.template('<a data-toggle="node" class="collapsed"><i class="fa fa-pie-chart"></i> <%=name%><div class="icon-group"><i class="fa fa-exclamation-triangle icon-error pull-right hide"></i><i class="fa fa-plus icon-expand pull-right"></i><i class="fa fa-minus icon-collapse hide pull-right"></i><i class="fa fa-spinner hide icon-loading fa-pulse"></i></div></a><div class="children"></div>')
+      }),
+      CustomNodeSuccessModel = Backbone.Model.extend({
+        fetchChildren: function (options) {
+          if (!options.err){
+            if (options.multiLvl){
```
Comment (Nelrohd) [inline test/unit/spec/node.js]: space before {

File: test/unit/spec/node.js [hunk 3]
```diff
@@ -1,115 +1,196 @@
...
-    customNodeView.$('a.collapsed[data-toggle=node] i').trigger(clickEvent);
-
-    expect(customNodeView.onExpand).toHaveBeenCalled();
-  });
+  var CustomNodeView = Marionette.NodeView.extend({
+        template: _.template('<a data-toggle="node" class="collapsed"><i class="fa fa-pie-chart"></i> <%=name%><div class="icon-group"><i class="fa fa-exclamation-triangle icon-error pull-right hide"></i><i class="fa fa-plus icon-expand pull-right"></i><i class="fa fa-minus icon-collapse hide pull-right"></i><i class="fa fa-spinner hide icon-loading fa-pulse"></i></div></a><div class="children"></div>')
+      }),
+      CustomNodeSuccessModel = Backbone.Model.extend({
+        fetchChildren: function (options) {
+          if (!options.err){
```
Comment (Nelrohd) [inline test/unit/spec/node.js]: space before {

File: test/unit/spec/node.js [hunk 4]
```diff
@@ -1,115 +1,196 @@
...
+            if (options.multiLvl){
+              this.set('children', new Backbone.Collection(multiLvlData));
+            } else {
+              this.set('children', new Backbone.Collection(mockData));
+            }
+          } else {
+            this.trigger('error');
+          }
+        },
+        deleteChildren: function () {
```
Comment (Nelrohd) [inline test/unit/spec/node.js]: no space before ()

File: test/unit/spec/node.js [hunk 5]
```diff
@@ -1,115 +1,196 @@
...
+          this.unset('children', {silent: true});
```
Comment (Nelrohd) [inline test/unit/spec/node.js]: space before silent and after true

File: test/unit/spec/node.js [hunk 6]
```diff
@@ -1,115 +1,196 @@
...
+  };
+
+// Tests
+  describe("NodeView", function () {
+
+    beforeEach(function () {
+      // Create object
+      var model = new CustomNodeSuccessModel({name: 'Root'});
+
+      this.customNodeView = new CustomNodeView({collectionType: CustomNodeSuccessCollection, model: model});
```
Comment (Nelrohd) [inline test/unit/spec/node.js]: space before collectionType

File: test/unit/spec/node.js [hunk 7]
```diff
@@ -1,115 +1,196 @@
...
+    reactOnCollapse: function () {},
+    reactOnSelect: function () {}
+  };
+
+// Tests
+  describe("NodeView", function () {
+
+    beforeEach(function () {
+      // Create object
+      var model = new CustomNodeSuccessModel({name: 'Root'});
```
Comment (Nelrohd) [inline test/unit/spec/node.js]: space before name

File: test/unit/spec/node.js [hunk 8]
```diff
@@ -1,115 +1,196 @@
...
-    customNodeView.delegateEvents();
-    customNodeView.$('a.collapsed[data-toggle=node] i').trigger(clickEvent);
-
-    expect(customNodeView.onExpand).toHaveBeenCalled();
-  });
+  var CustomNodeView = Marionette.NodeView.extend({
+        template: _.template('<a data-toggle="node" class="collapsed"><i class="fa fa-pie-chart"></i> <%=name%><div class="icon-group"><i class="fa fa-exclamation-triangle icon-error pull-right hide"></i><i class="fa fa-plus icon-expand pull-right"></i><i class="fa fa-minus icon-collapse hide pull-right"></i><i class="fa fa-spinner hide icon-loading fa-pulse"></i></div></a><div class="children"></div>')
+      }),
+      CustomNodeSuccessModel = Backbone.Model.extend({
+        fetchChildren: function (options) {
```
Comment (Nelrohd) [inline test/unit/spec/node.js]: no space before (

File: test/unit/spec/node.js [hunk 9]
```diff
@@ -1,115 +1,196 @@
...
+      });
+
+  var controller = {
+    reactOnExpand: function () {},
+    reactOnCollapse: function () {},
+    reactOnSelect: function () {}
+  };
+
+// Tests
+  describe("NodeView", function () {
```
Comment (Nelrohd) [inline test/unit/spec/node.js]: no space before ()

File: test/unit/spec/node.js [hunk 10]
```diff
@@ -1,115 +1,196 @@
...
+          this.unset('children', {silent: true});
+          this.trigger('delete:children');
+        }
+      }),
+      CustomNodeSuccessCollection = Backbone.Collection.extend({
+        model: CustomNodeSuccessModel
+      });
+
+  var controller = {
+    reactOnExpand: function () {},
```
Comment (Nelrohd) [inline test/unit/spec/node.js]: no space before

File: test/unit/spec/node.js [hunk 11]
```diff
@@ -1,115 +1,196 @@
...
+    reactOnCollapse: function () {},
```
Comment (Nelrohd) [inline test/unit/spec/node.js]: no space before

File: test/unit/spec/node.js [hunk 12]
```diff
@@ -1,115 +1,196 @@
...
+    reactOnSelect: function () {}
```
Comment (Nelrohd) [inline test/unit/spec/node.js]: no space before

File: test/unit/spec/node.js [hunk 13]
```diff
@@ -1,115 +1,196 @@
...
+
+    beforeEach(function () {
+      // Create object
+      var model = new CustomNodeSuccessModel({name: 'Root'});
+
+      this.customNodeView = new CustomNodeView({collectionType: CustomNodeSuccessCollection, model: model});
+      this.customNodeView.render();
+    });
+
+    afterEach(function () {
```
Comment (Nelrohd) [inline test/unit/spec/node.js]: no space before ()

File: test/unit/spec/node.js [hunk 14]
```diff
@@ -1,115 +1,196 @@
...
+      this.customNodeView.destroy();
+    });
+
+    it("should be rendered well", function () {
```
Comment (Nelrohd) [inline test/unit/spec/node.js]: no space before ()

File: test/unit/spec/node.js [hunk 15]
```diff
@@ -1,115 +1,196 @@
...
+      expect(this.customNodeView.$el.html()).toContain('Root');
+      expect(this.customNodeView.ui.iconError.hasClass('hide')).toBeTruthy();
+      expect(this.customNodeView.ui.iconLoading.hasClass('hide')).toBeTruthy();
+      expect(this.customNodeView.ui.iconCollapse.hasClass('hide')).toBeTruthy();
+      expect(this.customNodeView.ui.iconExpand.hasClass('hide')).toBeFalsy();
+    });
+
+    it("should hide the expand button when a model has no children", function () {
+      this.customNodeView.model.fetchChildren({ err: false, multiLvl: false});
```
Comment (Nelrohd) [inline test/unit/spec/node.js]: space after false

File: test/unit/spec/node.js [hunk 16]
```diff
@@ -1,115 +1,196 @@
...
+
+      expect(this.customNodeView.children.first().ui.iconExpand.hasClass('hide')).toBeTruthy();
+      expect(this.customNodeView.children.last().ui.iconExpand.hasClass('hide')).toBeFalsy();
+    });
+
+    it("should show the appropriate icons when loading models that have children", function () {
+      this.customNodeView.model.fetchChildren({ err: false, multiLvl: true});
```
Comment (Nelrohd) [inline test/unit/spec/node.js]: space after true

File: test/unit/spec/node.js [hunk 17]
```diff
@@ -1,115 +1,196 @@
...
+      expect(this.customNodeView.children.first().children.first().ui.iconCollapse.hasClass('hide')).toBeFalsy();
+
+      expect(this.customNodeView.children.first().children.last().ui.iconExpand.hasClass('hide')).toBeFalsy();
+      expect(this.customNodeView.children.first().children.last().ui.iconCollapse.hasClass('hide')).toBeTruthy();
+
+      expect(this.customNodeView.children.first().children.first().children.first().ui.iconExpand.hasClass('hide')).toBeFalsy();
+      expect(this.customNodeView.children.first().children.first().children.first().ui.iconCollapse.hasClass('hide')).toBeTruthy();
+    });
+
+    it("should trigger the expand event and call the associated callback", function () {
```
Comment (Nelrohd) [inline test/unit/spec/node.js]: no space before ()

File: test/unit/spec/node.js [hunk 18]
```diff
@@ -1,115 +1,196 @@
...
+      spyOn(controller, 'reactOnExpand');
+
+      this.customNodeView.bind('expand', controller.reactOnExpand, controller);
+      this.customNodeView.ui.iconExpand.trigger($.Event('click'));
+
+      expect(controller.reactOnExpand).toHaveBeenCalled();
+    });
+
+    it("should be expanded", function () {
```
Comment (Nelrohd) [inline test/unit/spec/node.js]: no space before ()

File: test/unit/spec/node.js [hunk 19]
```diff
@@ -1,115 +1,196 @@
...
+      expect(this.customNodeView.$el.html()).toContain('Root');
+      expect(this.customNodeView.$el.html()).toContain('child1');
+      expect(this.customNodeView.ui.iconError.hasClass('hide')).toBeTruthy();
+      expect(this.customNodeView.ui.iconLoading.hasClass('hide')).toBeTruthy();
+      expect(this.customNodeView.ui.iconExpand.hasClass('hide')).toBeTruthy();
+      expect(this.customNodeView.ui.iconCollapse.hasClass('hide')).toBeFalsy();
+      expect(this.customNodeView.isExpanding).toBeFalsy();
+    });
+
+    it("should trigger the collapse event and call the associated callback", function () {
```
Comment (Nelrohd) [inline test/unit/spec/node.js]: no space before ()

File: test/unit/spec/node.js [hunk 20]
```diff
@@ -1,115 +1,196 @@
...
+
+      expect(this.customNodeView.$el.html()).toContain('Root');
+      expect(this.customNodeView.$el.html()).not.toContain('child1');
+      expect(this.customNodeView.ui.iconError.hasClass('hide')).toBeTruthy();
+      expect(this.customNodeView.ui.iconLoading.hasClass('hide')).toBeTruthy();
+      expect(this.customNodeView.ui.iconCollapse.hasClass('hide')).toBeTruthy();
+      expect(this.customNodeView.ui.iconExpand.hasClass('hide')).toBeFalsy();
+    });
+
+    it("should be active after a click event on the node", function () {
```
Comment (Nelrohd) [inline test/unit/spec/node.js]: no space before ()

File: test/unit/spec/tree.js [hunk 21]
```diff
@@ -0,0 +1,125 @@
...
+    name: 'leaf2'
+  }];
+
+  var CustomNodeModel = Backbone.Model.extend({
+    fetchChildren: function(options) {
+      this.set('children', new CustomNodeCollection(otherData));
+    },
+
+    destroyChildren: function(options) {
+      this.unset('children', {silent: true});
```
Comment (Nelrohd) [inline test/unit/spec/tree.js]: space before silent and after true

File: test/unit/spec/tree.js [hunk 22]
```diff
@@ -0,0 +1,125 @@
...
+
+    onExpand: function(model) { model.fetchChildren(); },
+
+    onCollapse: function(model) { model.destroyChildren(); },
+
+    onSelect: function(model) { this.trigger('select', model); }
+  });
+
+  var controller = {
+    reactOnSelect: function () {},
```
Comment (Nelrohd) [inline test/unit/spec/tree.js]: no space before ()

File: test/unit/spec/tree.js [hunk 23]
```diff
@@ -0,0 +1,125 @@
...
+    reactOnCollapse: function () {}
```
Comment (Nelrohd) [inline test/unit/spec/tree.js]: no space before ()

File: test/unit/spec/tree.js [hunk 24]
```diff
@@ -0,0 +1,125 @@
...
+  };
+
+  describe("TreeView", function() {
+
+    beforeEach(function() {
+      this.customLayout = new CustomLayout();
+      this.customLayout.render();
+    });
+
+    it("should be well initialized", function () {
```
Comment (Nelrohd) [inline test/unit/spec/tree.js]: no space before ()

File: test/unit/spec/tree.js [hunk 25]
```diff
@@ -0,0 +1,125 @@
...
+      expect(this.customLayout).toBeDefined();
+      expect(this.customLayout.treeView.currentView instanceof Marionette.TreeView).toBeTruthy();
+      expect(this.customLayout.treeView.currentView.collection instanceof CustomNodeCollection).toBeTruthy();
+      expect(this.customLayout.treeView.currentView.collection.first() instanceof CustomNodeModel).toBeTruthy();
+      expect(this.customLayout.treeView.currentView.children.findByIndex(0) instanceof CustomNodeView).toBeTruthy();
+    });
+
+    it("should set the first node as selected/active after the view is showed", function () {
```
Comment (Nelrohd) [inline test/unit/spec/tree.js]: no space before ()

File: test/unit/spec/tree.js [hunk 26]
```diff
@@ -0,0 +1,125 @@
...
+      expect(this.customLayout.treeView.currentView.children.first().$el.hasClass('active')).toBeTruthy();
+      expect(this.customLayout.treeView.currentView.children.findByIndex(1).$el.hasClass('active')).toBeFalsy();
+      expect(this.customLayout.treeView.currentView.children.findByIndex(2).$el.hasClass('active')).toBeFalsy();
+    });
+
+    it("should change the selected/active node after a select event is triggered", function () {
```
Comment (Nelrohd) [inline test/unit/spec/tree.js]: no space before ()

File: test/unit/spec/tree.js [hunk 27]
```diff
@@ -0,0 +1,125 @@
...
+      expect(this.customLayout.treeView.currentView.children.findByIndex(2).$el.hasClass('active')).toBeFalsy();
+
+      this.customLayout.treeView.currentView.children.findByIndex(1).ui.toggle.trigger($.Event('click'));
+
+      expect(this.customLayout.treeView.currentView.children.findByIndex(0).$el.hasClass('active')).toBeFalsy();
+      expect(this.customLayout.treeView.currentView.children.findByIndex(1).$el.hasClass('active')).toBeTruthy();
+      expect(this.customLayout.treeView.currentView.children.findByIndex(2).$el.hasClass('active')).toBeFalsy();
+    });
+
+    it("should listen to the event bound to childviews that were added after the initial creation of the treeview", function () {
```
Comment (Nelrohd) [inline test/unit/spec/tree.js]: no space before ()

File: test/unit/spec/tree.js [hunk 28]
```diff
@@ -0,0 +1,125 @@
...
+
+      this.customLayout.treeView.currentView.collection.add({ name: 'newChild' });
+      this.customLayout.treeView.currentView.bind('collapse', controller.reactOnCollapse, controller);
+      expect(this.customLayout.treeView.currentView.children.last().model.get('name')).toEqual('newChild');
+
+      this.customLayout.treeView.currentView.children.last().ui.iconCollapse.trigger($.Event('click'));
+      expect(controller.reactOnCollapse).toHaveBeenCalled();
+    });
+
+    it("should trigger the select event, propagate it towards the treeview's controller and call it's associated callback", function () {
```
Comment (Nelrohd) [inline test/unit/spec/tree.js]: no space before ()

File: test/unit/spec/tree.js [hunk 29]
```diff
@@ -0,0 +1,125 @@
...
+
+      expect(this.customLayout.treeView.currentView.children.first().children.first().model.get('name')).toEqual('leaf1');
+      this.customLayout.treeView.currentView.children.first().children.first().ui.toggle.trigger($.Event('click'));
+
+      expect(controller.reactOnSelect).toHaveBeenCalledWith(this.customLayout.treeView.currentView.children.first().children.first().model);
+    });
+
+  });
+
+})();
```
Comment (Nelrohd) [inline test/unit/spec/tree.js]: add a new line for the end of file

**Comment by Nelrohd** [comment]
Fix all of these comments and you will get +1 :P

File: test/unit/spec/tree.js (hunk 29 above)
Comment (FredFaust) [inline test/unit/spec/tree.js]: what is the purpose of that new line ?

File: dist/marionette-tree-view.js (hunk 1 above)
Comment (FredFaust) [inline dist/marionette-tree-view.js]: removed parenthesis

File: test/unit/spec/tree.js (hunk 29 above)
Comment (FredFaust) [inline test/unit/spec/tree.js]: I guess this diff isn't outdated because I added something (a new line) after this code ?

**Comment by FredFaust** [comment]
1 failed :octopus:

**Pull request merged**